### PR TITLE
[TM READY] Ports NTSL from Yogstation

### PR DESCRIPTION
**Pull request by tf-4**
## About The Pull Request

Now test-merge ready!

## About NTSL

> The NT Scripting Language(NT Script, or NTSL) is a new piece of technology being pushed by the NT Tech Department to standardize programming and communication of all Nanotrasen-grade electronic devices. Its syntax is a mixture of PHP, C++, and JavaScript. Most implementations of NT Script are not object-oriented and do not allow the definition of classes, instead, rely on ROBUST™️ memory management system to store entities in a dynamic data tree. NT Script does allow the in-line definition of functions, however.

> It is important to note that NT Scripting should not be treated as BYOND coding. NT Scripting handles some things differently than BYOND does.

Basically, you can write scripts for telecoms to adjust what information it gives out and in what way. This will be handled by a new job, the Signal Technician (subordinate to the CE).
image

They'll be in charge of developing and running the scripts, as well as the general upkeep of comms throughout the shift.

It works with all languages! It works on all station channels! (Not CC, syndicate, etc. channels). It works with silicons! It works underwater (probably)!

How do they do this, you ask? Well, it's simple!(ish).

1. Get to the Telecommunications Control Room
2. Open up the Telecommunications Traffic Control Console
3. Authorize yourself access by clicking on Insert ID while holding your ID
4. [Set] the network to 'tcommsat' 
5. [Scan] for servers
6. Click on a server which radio channel you wish to alter
7. [Edit Code] to see the server's code
8. Type your code into the window
9. Press Compile

10. Change Signal Execution to ALWAYS
11. Congratulations, you're done!
12. Repeat steps 5 to 10 for each desired channel

You can see examples of the interface below!
screens

And here's an example of how you can edit comms, though your imagination is pretty much your limit here.
examples

This will all go on the wiki soon™️, but for now, Yogstation has a pretty comprehensive guide [here](https://wiki.yogstation.net/wiki/NT_Script), as well as a repository of scripts (that may or may not be in date) [here](https://wiki.yogstation.net/wiki/NTSL_Scripts)

For jannies:
You can blank all comms scripts at the press of a button
fun admin times

## Major differences from Yogstation NTSL

- All the 🤢 relative pathing 🤢 has become absolute
- ZephyrTFA did a wonderful job with language defines
- That's about it really, it's a very close port with only differences in code and none in functionality

## Todo

- [x] Chip away at the 200+ relative paths
- [x] Add Signal Technician job
- [x] Plasmaman SigTech clothing
- [x] Fix an issue with message contents getting lost (thanks ZephyrTFA)
- [x] SigTech ID trim visuals
- [ ] Scrub out the dirty single letter variables that I couldn't be bothered to do yet

## Test Merge issues

- [x] Sig Techs don't have proper engineering access
- [ ] IceBox ignored my map changes, for some reason
- [ ] Hallucinated radio messages don't use NTSL
- [ ] Door access on some maps, like Delta
- [ ] **Messages run through a script don't show in runechat**
- [ ] The captain doesn't get ACCESS_TCOM_ADMIN heehee
- [ ] Achievements are possibly borked

## Changelog
:cl:
add: Nanotrasen have upgraded their telecoms setups, allowing for the use of NT Script for far more control over what goes over the airwaves!
add: Now hiring: Signal Technicians! As a SigTech, your job will be maintaining telecoms as well as developing scripts to ~~annoy~~ enrich the experience of the crew's earballs!
remove: Emoting over the radio has ceased to be, though it's expected to return at some point in the future.
/:cl:

### CODE REVIEW

**Comment by Gandalf2k15** [comment]
Mad man

**Comment by NullDagaf** [comment]
This will only end in an autistic blaze of glory.

**Comment by gonenoculer5** [comment]
Could we actually get some information on what this does in the changelog? This doesnt tell us a whole lot.
> add: NTSL

**Comment by tf-4** [comment]
> 
> 
> Could we actually get some information on what this does in the changelog? This doesnt tell us a whole lot.
> 
> > add: NTSL

> [HEAVY WIP]

**Comment by NullDagaf** [comment]
Oh god, it works?

**Comment by Elli-Skala** [comment]
problem found, it counts all of comms as not speaking common leading to translator circuits translating everything said over radio

**Comment by tf-4** [comment]
> problem found, it counts all of comms as not speaking common leading to translator circuits translating everything said over radio

Pain

**Comment by werewolfboyfriend** [comment]
I tried signal tech last night on tram and found that I had no access to get to tcomms (the doors leading from the eng lobby to the room in front of the SM) and the upper blue door in tcomms didn't have sigtech access but the bottom one did. Just to make sure that both of those are known.
Cool, otherwise. Might be neat to have events that happen only when a sigtech is on shift to give them stuff to do or fix.

**Comment by tf-4** [comment]
> 
> 
> I tried signal tech last night on tram and found that I had no access to get to tcomms (the doors leading from the eng lobby to the room in front of the SM) and the upper blue door in tcomms didn't have sigtech access but the bottom one did. Just to make sure that both of those are known.
> Cool, otherwise. Might be neat to have events that happen only when a sigtech is on shift to give them stuff to do or fix.

yeah this was fixed a while ago but the server likely wasn't updated. Should be fine now AFAIK

**Comment by ZephyrTFA** [comment]
> 
> 
> problem found, it counts all of comms as not speaking common leading to translator circuits translating everything said over radio

I'll look into this as language handling is my part of NTSL.

**Comment by ZephyrTFA** [comment]
In bug testing that issue, I found a MUCH LARGER ISSUE, thank you for the report @Elli-Skala

**Comment by YakumoChen** [comment]
I liked NTSL but is this more secure than it was before it was removed from TG code? iirc it was removed because it was a major server security problem.

The job also sort of feels like it's a little too overreliant on having a wiki/database of scripts / having intimate knowledge of actual code. Is there an in-game way to get some basic knowledge of how this works or access some kind of basic database of scripts?

**Comment by tf-4** [comment]
> I liked NTSL but is this more secure than it was before it was removed from TG code? iirc it was removed because it was a major server security problem.
> 
> The job also sort of feels like it's a little too overreliant on having a wiki/database of scripts / having intimate knowledge of actual code. Is there an in-game way to get some basic knowledge of how this works or access some kind of basic database of scripts?

I'll add a wiki book linking to our wiki for NTSL if this gets fullmerged. I don't want to make a wiki page for something that may still get scrapped

File: modular_skyrat/modules/ntsl/code/datums/achievements/achievements.dm [hunk 1]
```diff
@@ -0,0 +1,9 @@
+/datum/award/achievement/engineering/Poly_silent
+	name = "Silence Bird!"
+	desc = "As a signal technician, create a script that mutes poly"
+	database_id = MEDAL_POLYSILENT
+
+/datum/award/achievement/engineering/Poly_loud
+	name = "Embrace the Bird!"
+	desc = "As a signal technician, create a script that makes poly LOUD"
+	database_id = MEDAL_POLYLOUD
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/ntsl/code/datums/achievements/achievements.dm]:
```suggestion
/datum/award/achievement/jobs/sig_tech_poly_silent
	name = "Silence Bird!"
	desc = "As a signal technician, create a script that mutes poly"
	database_id = MEDAL_POLYSILENT

/datum/award/achievement/jobs/sig_tech_poly_loud
	name = "Embrace the Bird!"
	desc = "As a signal technician, create a script that makes poly LOUD"
	database_id = MEDAL_POLYLOUD
```
This, alongside my next suggested change, should fix the achievements issues.

File: modular_skyrat/modules/ntsl/code/game/machinery/telecomms/machines/server.dm [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+			if(signal.data["reject"] == 1)
+				signal.data["name"] = ""
+				signal.data["reject"] = 0
+				Compiler.Run(signal)
+				if(signal.data["reject"] == 0)
+					usr.client.give_award(/datum/award/achievement/engineering/Poly_silent, usr)
+			else
+				for(var/sample in signal.data["spans"])
+					if(sample == SPAN_COMMAND)
+						usr.client.give_award(/datum/award/achievement/engineering/Poly_loud, usr)
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/ntsl/code/game/machinery/telecomms/machines/server.dm]:
```suggestion
					usr.client.give_award(/datum/award/achievement/jobs/sig_tech_poly_silent, usr)
			else
				for(var/sample in signal.data["spans"])
					if(sample == SPAN_COMMAND)
						usr.client.give_award(/datum/award/achievement/jobs/sig_tech_poly_loud, usr)
```
Second part to fix the achievements.

**Comment by tf-4** [comment]
Too many inherent balance and RP issues, as well as waning interest in working on it. If anyone else wants to pick it up, feel absolutely free

**Pull request closed without merging**